source/develop/release-management/features/storage/multipath-events.html.md

```diff
@@ -0,0 +1,201 @@
+### New vdsm API
+
+New property in 'HostStats': 'MultipathAlertsMap' which is a mapping of faulty paths
+indexed by the device mapper GUID.
+
+
+#### Multipath Alerts Map
+
+The key of the map is the device mapper GUID.
+The value is a MultipathAlert that contains:
+- List of faulty paths.
+- Number of valid paths
+
+### Vdsm multipath event listener
+
+Introduce a new Vdsm internal service for listening to multipath events:
+
+- Code interested in receiving multipath events will subscribe a callback to receive the event
+- When an event is received, all subscribers will receive the events (in another thread).
+- If nobody is registered for events, the events are dropped
+- The multipath event monitor will be started during storage initialization
+
+
+The service will be implemented using 'pyudev', listening to these events:
+- action=change,DM_ACTION=PATH_FAILED
+- action=change,DM_ACTION=PATH_REINSTATED
+- action=remove (for multipath devices)
+
+We don't know yet how to detect multipath devices events efficiently for detecting removed devices.
+
+
+- The user can disable this feature in a specific host in Vdsm configuration file
+
+In the future, the monitor can be used for other code, for example maintaining LVM filter.
+
+### Vdsm multipath health monitor
+
+To report multipath health status to engine, we will add a multipath health monitor.
+
+- The monitor subscribes to multipath events during startup
+- The monitor builds a map of all the faulty paths during startup.
+- The monitor updates the data set of the faulty paths when its callback is called:
+    - PATH_FAILED - Add the faulty path to the map according to the device GUID
+    - PATH_REINSTATED - Remove the faulty path to the map according to the device GUID
+    - Multipath device removed  - remove the device from the map
+- Multipath health status is provided to engine via 'Host.getStats' verb.
+
+
+### New engine API
+
+New audit logs are added:
+- Faulty multipath paths on host "HOST_NAME" on devices: "GUID", "GUID" ...
+- Devices without valid paths on host "HOST_NAME" : "GUID", "GUID" ...
+- No faulty multipath paths on host "HOST_NAME"
+
+## Benefit to oVirt
+
+- The user will be able to detect storage issues faster.
+
+
+## User Experience
+
+New audit logs are added with information on the device path status.
+The logs are per host.
+Since the length of the information available in the audit log is limited,
+the full information of the failing paths will be available in the logs.
+The events are not generated on "real-time", but are based on polling of
+the statistics. Only if a change in the data received is detected, a new
+audit log will be created
+
+If a single LUN goes down on all the hosts, the user will get an alarm
+for each host in the DC.
+
+If a single path goes down on one of the hosts, the user will get an alarm
+only for this host.
+
+
+## Installation/Upgrade
+
+This feature does not affect engine or Vdsm installation or upgrades
+
+
+## User work-flows
```

Comment:
workflow is single word.
http://www.dictionary.com/browse/workflow?s=t

Reply:
done